File: regions/system/library/python/test_helper/src/k1s0_test_helper/assertion.py
"""テスト用アサーションヘルパー。"""

from __future__ import annotations

import json
from typing import Any


class AssertionHelper:
    """テスト用アサーションヘルパー。"""

    @staticmethod
    def assert_json_contains(actual: Any, expected: Any) -> None:
        """JSON 部分一致アサーション。"""
        if isinstance(actual, str):
            actual = json.loads(actual)
        if isinstance(expected, str):
            expected = json.loads(expected)
        if not _json_contains(actual, expected):
            raise AssertionError(
                f"JSON partial match failed.\n"
                f"Actual: {json.dumps(actual)}\n"
                f"Expected: {json.dumps(expected)}"
            )

    @staticmethod
    def assert_event_emitted(events: list[dict[str, Any]], event_type: str) -> None:
        """イベント一覧に指定タイプのイベントが含まれるか検証する。"""
        found = any(e.get("type") == event_type for e in events)
        if not found:
            raise AssertionError(f"Event '{event_type}' not found in events")


class AssertionError(Exception):
    """アサーションエラー型。"""


def _json_contains(actual: Any, expected: Any) -> bool:
    if isinstance(expected, dict):
        if not isinstance(actual, dict):
            return False
        return all(
            k in actual and _json_contains(actual[k], v)
            for k, v in expected.items()
        )
    if isinstance(expected, list):
        if not isinstance(actual, list):
            return False
        return all(
            any(_json_contains(av, ev) for av in actual) for ev in expected
        )
    return actual == expected

File: regions/system/library/dart/test_helper/lib/src/assertion.dart
import 'dart:convert';

/// テスト用アサーションヘルパー。
class AssertionHelper {
  /// JSON 部分一致アサーション。
  static void assertJsonContains(dynamic actual, dynamic expected) {
    final actualObj = actual is String ? jsonDecode(actual) : actual;
    final expectedObj = expected is String ? jsonDecode(expected) : expected;
    if (!_jsonContains(actualObj, expectedObj)) {
      throw AssertionError(
        'JSON partial match failed.\n'
        'Actual: ${jsonEncode(actualObj)}\n'
        'Expected: ${jsonEncode(expectedObj)}',
      );
    }
  }

  /// イベント一覧に指定タイプのイベントが含まれるか検証する。
  static void assertEventEmitted(List<dynamic> events, String eventType) {
    final found = events.any((e) =>
        e is Map<String, dynamic> && e['type'] == eventType);
    if (!found) {
      throw AssertionError("Event '$eventType' not found in events");
    }
  }

  static bool _jsonContains(dynamic actual, dynamic expected) {
    if (expected is Map<String, dynamic>) {
      if (actual is! Map<String, dynamic>) return false;
      return expected.entries
          .every((e) => actual.containsKey(e.key) &&
              _jsonContains(actual[e.key], e.value));
    }
    if (expected is List) {
      if (actual is! List) return false;
      return expected.every((ev) => actual.any((av) => _jsonContains(av, ev)));
    }
    return actual == expected;
  }
}

/// アサーションエラー型。
class AssertionError implements Exception {
  final String message;
  const AssertionError(this.message);

  @override
  String toString() => message;
}
